sdk/src/asset_handlers/riff_io.rs

```diff
@@ -198,46 +366,94 @@ impl AssetIO for RiffIO {
     }
 
     fn save_cai_store(&self, asset_path: &std::path::Path, store_bytes: &[u8]) -> Result<()> {
-        let asset = std::fs::read(asset_path)?;
-        let mut chunk_reader = Cursor::new(asset);
+        let mut input_stream = File::open(asset_path)?;
+
+        let mut output_stream = std::fs::OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open(asset_path)
+            .map_err(Error::IoError)?;
+
+        self.write_cai(&mut input_stream, &mut output_stream, store_bytes)
+    }
+
+    fn get_object_locations(
+        &self,
+        asset_path: &std::path::Path,
+    ) -> Result<Vec<HashObjectPositions>> {
+        let mut f = std::fs::File::open(asset_path).map_err(|_err| Error::EmbeddingError)?;
+
+        self.get_object_locations_from_stream(&mut f)
+    }
+
+    fn remove_cai_store(&self, asset_path: &Path) -> Result<()> {
+        self.save_cai_store(asset_path, &[])
+    }
+
+    fn remote_ref_writer_ref(&self) -> Option<&dyn RemoteRefEmbed> {
+        Some(self)
+    }
 
-        let top_level_chunks = Chunk::read(&mut chunk_reader, 0)?;
+    fn supported_types(&self) -> &[&str] {
+        &SUPPORTED_TYPES
+    }
+}
+
+impl CAIWriter for RiffIO {
+    fn write_cai(
+        &self,
+        input_stream: &mut dyn CAIRead,
+        output_stream: &mut dyn CAIReadWrite,
+        store_bytes: &[u8],
+    ) -> Result<()> {
+        let top_level_chunks = {
+            let mut reader = CAIReadWrapper {
+                reader: input_stream,
+            };
+            Chunk::read(&mut reader, 0)?
+        };
 
         if top_level_chunks.id() != RIFF_ID {
             return Err(Error::InvalidAsset("Invalid RIFF format".to_string()));
         }
 
+        let mut reader = CAIReadWrapper {
+            reader: input_stream,
+        };
+
         // replace/add manifest in memory
         let new_contents = inject_c2pa(
             &top_level_chunks,
-            &mut chunk_reader,
+            &mut reader,
             store_bytes,
+            None,
             &self.riff_format,
         )?;
 
+        let mut writer = CAIReadWriteWrapper {
+            reader_writer: output_stream,
+        };
+
         // save contents
-        let mut output = OpenOptions::new()
-            .read(true)
-            .write(true)
-            .open(asset_path)
-            .map_err(Error::IoError)?;
-        match new_contents.write(&mut output) {
-            Ok(_) => Ok(()),
-            Err(e) => Err(Error::IoError(e)),
-        }
+        new_contents
+            .write(&mut writer)
+            .map_err(|_e| Error::EmbeddingError)?;
```

Comment:
This makes it easier to tell where the error happened, but we also lose track of the actual error for things like permissions and such. I'm not sure what our overall policy is for this kind of thing

Reply:
The mapping here is because of the return type.  Let me see what I can do about this.